sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala

```diff
@@ -36,17 +36,18 @@ import org.apache.spark.sql.util.SchemaUtils
  * Replaces [[UnresolvedRelation]]s if the plan is for direct query on files.
  */
 class ResolveSQLOnFile(sparkSession: SparkSession) extends Rule[LogicalPlan] {
+
   private def maybeSQLFile(u: UnresolvedRelation): Boolean = {
-    sparkSession.sessionState.conf.runSQLonFile && u.tableIdentifier.database.isDefined
+    sparkSession.sessionState.conf.runSQLonFile && u.multipartIdentifier.size == 2
```

Comment:
This is a separate problem, but could you also file an issue to fix this rule? This relies on rule ordering for correctness because it will convert any 2-part identifier to a SQL-on-file plan if the namespace is also a DataSource name. This should additionally check that the path either looks like a path (contains `/`) or exists in some file system.

Reply:
Sounds good.

What should be the expected behavior for this test in sql.SQLQuerySuite."run sql directly on files"?
```
    e = intercept[AnalysisException] {
      sql("select * from json.invalid_file")
    }
    assert(e.message.contains("Path does not exist"))
```

Reply:
I think it should be that the table could not be found. If there were some way to determine that it is a path, like checking for `/`, then that message would be correct. So it could be updated to this:

```scala
    e = intercept[AnalysisException] {
      sql("select * from json.`/invalid_file`")
    }
    assert(e.message.contains("Path does not exist"))
```